ext/node/zlib/mod.rs

```diff
@@ -0,0 +1,474 @@
+          && self.mode == GUNZIP
+          && self.err == Z_STREAM_END
+          && unsafe { *self.strm.next_in as u8 } != 0x00
+        {
+          self.err = unsafe { libz_sys::inflateReset(&mut self.strm) };
+          self.err = unsafe { libz_sys::inflate(&mut self.strm, self.flush) };
+        }
+      }
+      _ => {}
+    }
+    if self.err == Z_BUF_ERROR
+      && !(self.strm.avail_out != 0 && self.flush == Z_FINISH)
+    {
+      self.err = Z_OK;
+    };
+    self.write_in_progress = false;
+    Ok(())
+  }
+}
+
+struct Zlib {
+  inner: RefCell<ZlibInner>,
+}
+
+impl deno_core::Resource for Zlib {
+  fn name(&self) -> Cow<str> {
+    "zlib".into()
+  }
+}
+
+const NONE: i32 = 0;
+const DEFLATE: i32 = 1;
+const INFLATE: i32 = 2;
+const GZIP: i32 = 3;
+const GUNZIP: i32 = 4;
+const DEFLATERAW: i32 = 5;
+const INFLATERAW: i32 = 6;
+const UNZIP: i32 = 7;
+
+#[op]
+pub fn op_zlib_new(state: &mut OpState, mode: i32) -> Result<u32, AnyError> {
+  if mode < DEFLATE || mode > UNZIP {
+    return Err(type_error("Bad argument"));
+  }
+  let strm: libz_sys::z_stream = libz_sys::z_stream {
+    next_in: std::ptr::null_mut(),
+    avail_in: 0,
+    total_in: 0,
+    next_out: std::ptr::null_mut(),
+    avail_out: 0,
+    total_out: 0,
+    msg: std::ptr::null_mut(),
+    state: std::ptr::null_mut(),
+    zalloc: alloc::zalloc,
+    zfree: alloc::zfree,
+    opaque: 0 as libz_sys::voidpf,
+    data_type: 0,
+    adler: 0,
+    reserved: 0,
+  };
+  let inner = ZlibInner {
+    strm,
+    mode,
+    dictionary: None,
+    err: 0,
+    flush: 0,
+    init_done: false,
+    level: 0,
+    mem_level: 0,
+    strategy: 0,
+    window_bits: 0,
+    write_in_progress: false,
+    pending_close: false,
+    gzib_id_bytes_read: 0,
+  };
+
+  Ok(state.resource_table.add(Zlib {
+    inner: RefCell::new(inner),
+  }))
+}
+
+#[op]
+pub fn op_zlib_close(state: &mut OpState, handle: u32) -> Result<(), AnyError> {
+  let resource = state
+    .resource_table
+    .get::<Zlib>(handle)
+    .map_err(|_| bad_resource_id())?;
+
+  let mut zlib = resource.inner.borrow_mut();
+  if zlib.write_in_progress {
+    zlib.pending_close = true;
+    return Ok(());
+  }
+
+  zlib.pending_close = false;
+  check(zlib.init_done, "close before init")?;
+  check(zlib.mode <= UNZIP, "invalid mode")?;
+
+  if zlib.mode == DEFLATE || zlib.mode == GZIP || zlib.mode == DEFLATERAW {
+    unsafe { libz_sys::deflateEnd(&mut zlib.strm) };
+  } else if zlib.mode == INFLATE
+    || zlib.mode == GUNZIP
+    || zlib.mode == INFLATERAW
+    || zlib.mode == UNZIP
+  {
+    unsafe { libz_sys::inflateEnd(&mut zlib.strm) };
+  }
+
+  zlib.mode = NONE;
+
+  Ok(())
+}
+
+#[op]
+pub fn op_zlib_write_async(
+  state: Rc<RefCell<OpState>>,
+  handle: u32,
+  flush: i32,
+  input: &[u8],
+  in_off: u32,
+  in_len: u32,
+  out: &mut [u8],
+  out_off: u32,
+  out_len: u32,
+) -> Result<
+  impl Future<Output = Result<(i32, u32, u32), AnyError>> + 'static,
+  AnyError,
+> {
+  let resource = state
```

Comment:
This particular pattern is pretty common in this file -- could we just have a single function to get a `Zlib` from `state`?

Reply:
resource needs to live longer than `zlib`. `Rc<Zlib>` will be dropped early while a local borrow is returned. We could clone `resource.inner` by making it an `Rc` but is that worth it?

Reply:
Sorry, I wasn't clear in my comment -- we could have a function that returns the `Rc`:

```rust
fn zlib(state: &mut OpState, handle: u32) -> Result<Rc<Zlib>, AnyError> {
  state
    .resource_table
    .get::<Zlib>(handle)
    .map_err(|_| bad_resource_id())
}
```

and then borrow from that:

```rust
#[op]
pub fn op_zlib_close_if_pending(
  state: &mut OpState,
  handle: u32,
) -> Result<(), AnyError> {
  let resource = zlib(state, handle)?;
  ... resource.inner.borrow_mut(); 
```

This is mainly because it's a complex bit of code and getting the boilerplate out of the way helps confirm correctness.